Fix line end bugs in Vim mode

sijieamoy opened the pull request:
## TLDR
Vim mode bug of handling line end when the next and last line is empty.

## Dive Deeper
- See https://github.com/google-gemini/gemini-cli/issues/5326 and https://github.com/google-gemini/gemini-cli/issues/5322
- Missing test cases for both

## Reviewer Test Plan
1. Enable Vim mode
2. Enter "hello world" followed by an empty line
3. Position the cursor on "w" then:
  1. `w` once;
  2.  `e` twice
4. In both cases, the cursor should move to the beginning of the empty line.

## Linked issues / bugs
- Fixes https://github.com/google-gemini/gemini-cli/issues/5326 
- Fixes https://github.com/google-gemini/gemini-cli/issues/5322

jacob314 commented on packages/cli/src/ui/components/shared/vim-buffer-actions.ts:
```diff
@@ -674,7 +765,64 @@ export function handleVimAction(
+          const currentLine = lines[cursorRow] || '';
+          const atEndOfWord =
+            cursorCol < currentLine.length &&
+            /\w/.test(currentLine[cursorCol]) &&
```
you need to use the codepoint helpers for all this logic to robustly handle unicode characters

jacob314 commented on packages/cli/src/ui/components/shared/vim-buffer-actions.ts:
```diff
@@ -18,6 +18,46 @@ import {
 
+// Helper function to handle word movement across empty lines.
+function findNextWordAcrossLines(
+  getText: () => string,
```
consider rewriting this logic to not require getText(). Same for the rest of the file. that will make it more robust as less conversion between lines and rows and columns would be needed. gemini cli may be able to handle that correctly with a oneshot.

sijieamoy replied:
Done.

jacob314 commented on packages/cli/src/ui/components/shared/text-buffer.ts:
```diff
@@ -33,143 +33,229 @@ function isWordChar(ch: string | undefined): boolean {
-): number => {
-  let i = currentOffset;
+// Helper functions for line-based word navigation
+export const isWordCharStrict = (char: string): boolean => /\w/.test(char);
```
does this have to be different from isWordChar in text-buffer.ts?
If not, would be ideal to converge.

sijieamoy replied:
`isWordCharStrict` is specifically for the vim behavior. An example is "hello@example.com". `isWordChar` will treat it as three words, "hello", "@example", ".com" when navigating with `alt+arrow`, but `isWordCharStrict` will see "hello", "@", "example", ".", "com" with Vim's word movement (`w`). So one is for the normal OS word nav and the other for the vim-specific logic.

jacob314 commented on packages/cli/src/ui/components/shared/text-buffer.ts:
```diff
@@ -33,143 +33,229 @@ function isWordChar(ch: string | undefined): boolean {
+  line: string,
+  col: number,
+): number | null => {
+  const chars = [...line]; // Handle Unicode properly
```
here and elsewhere please use the existing toCodePoints
helper in textUtils. which does the same thing to keep how utf-32 code point handling consistent.

jacob314 reviewed (commented):
Looks good! Please also add a couple of tests that verify that these methods behave correctly given UTF32 characters.

sijieamoy commented:
> Looks good! Please also add a couple of tests that verify that these methods behave correctly given UTF32 characters.

Addressed comments above and added a few UTF-32-related tests, specifically precomposed characters, combining characters, non-Latin characters, unicode symbols (excluding emojis as the input component doesn't seem to support emoji anyway).

jacob314 commented on packages/cli/src/ui/components/shared/text-buffer.test.ts:
```diff
@@ -1591,3 +1595,94 @@ describe('textBufferReducer vim operations', () => {
+
+    it('should handle Chinese characters', () => {
+      const result = findNextWordStartInLine('hello 你好 world', 0);
+      expect(result).toBe(6); // Start of Chinese word
```
would be good to also test if the first word was Chinese. Same above as that would probably stress improper unicode character handling more.

sijieamoy replied:
added

jacob314 commented on packages/cli/src/ui/components/shared/text-buffer.ts:
```diff
@@ -34,16 +34,41 @@ function isWordChar(ch: string | undefined): boolean {
 // Helper functions for line-based word navigation
-export const isWordCharStrict = (char: string): boolean => /\w/.test(char);
+export const isWordCharStrict = (char: string): boolean =>
+  /[\w\p{L}\p{N}]/u.test(char);
```
this is a tricky regex. 
Maybe add a comment that it
matches a single character that is any Unicode letter, any Unicode number, or an underscore

sijieamoy replied:
added

jacob314 commented on packages/cli/src/ui/components/shared/text-buffer.ts:
```diff
@@ -34,16 +34,41 @@ function isWordChar(ch: string | undefined): boolean {
+  return 'other';
+};
+
+// Check if two characters are from different scripts (indicating word boundary)
```
this has gotten really complex. could be good to audit what vim typically does and align with that. would be good to do the simplest that is sufficient.

sijieamoy replied:
Mind if I file a ticket as a reminder for later? The user-facing behaviors are consistent with Vim on these supported cases for now.

jacob314 commented on packages/cli/src/ui/components/shared/text-buffer.ts:
```diff
@@ -109,37 +151,95 @@ export const findPrevWordStartInLine = (
+        lastBaseCharPos = i;
+      }
+
+      // Check if next character is from a different script (word boundary)
```
are these cases really useful? if so please make sure there are tests that capture them

sijieamoy replied:
They are useful, and already have tests for them (starting at line 909 of vim-buffer-actions.test.ts).

jacob314 reviewed (commented):
The input prompt should support emojis. please tell me more about what is broken for them. seems like you may have caught a few more complex text cases we should handle better. would like to discuss a bit how we handle them correctly at a more fundamental level in text-buffer rather than just patching in the vim support.

sijieamoy commented on packages/cli/src/ui/components/shared/vim-buffer-actions.test.ts:
```diff
@@ -140,6 +140,25 @@ describe('vim-buffer-actions', () => {
         expect(result.cursorCol).toBe(0);
       });
+
+      it('should skip over combining marks to avoid cursor disappearing', () => {
```
What's the suggested action here? During manual testing the only case I found in the non-vim-mode is a word with mixed scripts e.g. "中文test英文word", `alt`+arrow doesn't detect the language boundaries.
